### DISCUSSION TITLE
Enable customizable User-Agent header in outgoing data requests (e.g. by make BuildVersion customizable in ini file)

### DISCUSSION THREAD

**Discussion by zhichli** [grafana/grafana#58714] [General]
- **What we are trying to achieve:**
We want to send custom value in `User-Agent` header of outgoing requests to data source endpoint, so that the data source endpoint can differentiate incoming requests based on the `User-Agent` header e.g. `User-Agent: Grafana/9.2.0` (self-host Grafana) vs `User-Agent: Grafana/build-1234 (a managed Grafana version)`, which is useful for business insights for data source service provider.
- **Current limitation:**
Currently, Grafana set `User-Agent` header via `BuildVersion` (in [http_client_provider.go](https://github.com/grafana/grafana/blob/dd0d034796c1c0a3b0ed5aa2924e602bb2d77a58/pkg/infra/httpclient/httpclientprovider/http_client_provider.go#L22) and [ds_proxy.go](https://github.com/grafana/grafana/blob/dd0d034796c1c0a3b0ed5aa2924e602bb2d77a58/pkg/api/pluginproxy/ds_proxy.go#L228)), which is not customizable given a pre-built image.
- **Feature request:**
It would help if the `BuildVersion` or `User-Agent` header value can be customized via `.ini` configuration

**Comment by wbrowne** [comment]
@grafana/plugins-platform-backend

**Comment by marefr** [comment]
@zhichli thanks. Just to get a better understanding, does it have to be the user agent header or could it be some other header containing the `build-1234`?

**Reply by zhichli** [reply]
@marefr thanks for the response. We prefer using `User-Agent` header since it's a widely-adopted one and already been used by our partner's data source endpoint to create business insights upon.
Our ultimate goal is to configure Grafana at startup using custom build version and let it flow to data source service endpoint where the `User-Agent` header seems the best transmission option. @leandro-deveikis

**Reply by marefr** [reply]
@zhichli sorry for slow response. Got it. Then your feature request sounds good to me. Just an FYI this will only work within Grafana and for core/builtin datasources. We cannot currently control/enforce the potential `User-Agent` set by external plugins, e.g. for azure-data-explorer-datasource there's [no user agent set](https://github.com/grafana/azure-data-explorer-datasource/blob/f4dd344da2339a6ae86180cd3959c8c71bd29b47/pkg/azuredx/client/client.go#L42-L51) 

Are you up for working on this? 

Guess next step would be to convert this to a feature request issue.

My suggestion would be to 
- introduce a new setting `user_agent` under `[dataproxy]` [configuration section](https://github.com/grafana/grafana/blob/a589929c92ec80d880ed6790f4c70af31a524e89/conf/defaults.ini#L172)
- introduce a new field `cfg.DataProxyUserAgent` or method in https://github.com/grafana/grafana/blob/main/pkg/setting/setting_data_proxy.go. If `[dataproxy].user_agent` set that will be the full `User-Agent`, if not set to the current, .e.g. `Grafana/<build version>'. With this we don't need to keep the user agent generation in multiple places.
- Use `cfg.DataProxyUserAgent` in [http_client_provider.go](https://github.com/grafana/grafana/blob/dd0d034796c1c0a3b0ed5aa2924e602bb2d77a58/pkg/infra/httpclient/httpclientprovider/http_client_provider.go#L22) and [ds_proxy.go](https://github.com/grafana/grafana/blob/dd0d034796c1c0a3b0ed5aa2924e602bb2d77a58/pkg/api/pluginproxy/ds_proxy.go#L228)

**Reply by zhichli** [reply]
@marefr I've created a #63769 based on your suggestion (Thanks for it!). Could you please review it at your convenience?

**Reply by marefr** [reply]
Implemented via https://github.com/grafana/grafana/issues/63767 so locking this discussion
